x/exchange/keeper/payments.go

```diff
@@ -0,0 +1,481 @@
+package keeper
+
+import (
+	"errors"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/quarantine"
+
+	"github.com/provenance-io/provenance/internal/antewrapper"
+	"github.com/provenance-io/provenance/x/exchange"
+)
+
+// paymentExists returns true if there's a payment in the store with the given source and external id.
+func paymentExists(store sdk.KVStore, source, externalID string) bool {
+	sourceAddr, err := sdk.AccAddressFromBech32(source)
+	if err != nil {
+		return false
+	}
+	return store.Has(MakeKeyPayment(sourceAddr, externalID))
+}
+
+// parsePaymentStoreValue converts a payment store value into the Payment object.
+func (k Keeper) parsePaymentStoreValue(value []byte) (*exchange.Payment, error) {
+	if len(value) == 0 {
+		return nil, nil
+	}
+
+	var payment exchange.Payment
+	err := k.cdc.Unmarshal(value, &payment)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal payment: %w", err)
+	}
+	return &payment, nil
```

Comment:
By returning `nil, nil` when the input byte slice is empty, it allows the caller to easily decide how to handle the case when there isn't a payment. If the caller needs to specially handle the case when there isn't a payment, they simply need to check the result against `nil`.

If I were to return an error when the input byte slice is empty, it's significantly more difficult for the caller to handle the case when there wasn't a payment because they'd then have to rely on parsing the content of the returned error.

Reply:
I update the function comment to include a description of this behavior.